README.md
# The Graylog CLI dashboard

This is a [Graylog](http://www.graylog.org/) stream dashboard that runs in your shell. It is meant to be a perfect companion for example
when performing a release on the main monitor and having the dashboard on the second monitor to keep an eye on errors and exceptions on
your platform.

![screencast](https://github.com/Graylog2/cli-dashboard/blob/master/screencast.gif)

The most important library we use is [blessed-contrib](https://github.com/yaronn/blessed-contrib)
by [Yaron Naveh](https://twitter.com/YaronNaveh). Great job on that!

**The dashboard should run on Linux and OSX.** Please report any issues in the [issue tracker](https://github.com/Graylog2/cli-dashboard/issues).

## Installation

Install the dashboard: (You'll have to have [node.js](http://nodejs.org/download/) > 4.0 installed.
Protip: It's in [homebrew](http://brew.sh) if you are on OSX.)

    $ npm install graylog-cli-dashboard -g

View the options:

    $ graylog-cli-dashboard --help
    Usage: graylog-dashboard.js <command> [options]

    Options:
      --stream-title    Graylog Stream Title
      --server-url      Full Graylog REST API URL
      --poll-interval   How often (in ms) to poll the Graylog server [default: 1000]
      --username        Graylog username
      --password        Graylog password
      --cred-file-path  Path to an optional credentials file
                                   [default: "/Users/user/.graylog_dashboard"]
      --help            Show help                                          [boolean]

Full example:

    $ graylog-dashboard --stream-title "Production Webservers" --host http://graylog.example.org:12900

Make sure to run this with a recent version (>= 4) of node.js.

#### Credentials File

If you wish, you may place a YAML file containing any of the above options at a path of your choosing.
The default is `~/.graylog_dashboard`. For example:

    username: lennart
    password: sEcReT

If you do not provide this file, and do not provide required data via command options, you will be prompted.

#### Specifying a Stream

Choose one from the list on the right side of the screen, or specify a `--stream-title` in the command.

#### If the graylog-dashboard executable cannot be found

You must have the `graylog-dashboard` executable that gets installed by NPM in your `PATH`. This is usually the case after installing
NPM in the first place but the output of `npm install` should show you where it was installed if you cannot find it:

    /usr/local/share/npm/bin/graylog-dashboard -> /usr/local/share/npm/lib/node_modules/graylog-cli-dashboard/graylog-dashboard.js
    # In this OSX example graylog-dashboard has been linked to the graylog-dashboard.js file in /usr/local/share/npm/...
    # Execute /usr/local/share/npm/bin/graylog-dashboard to try it out

### The Dashboard in the Wild

Because this stuff looks like it is from space we suspect that people like to show it. **Send us a photo of your dashboard running somewhere**
and we are more than happy to publish it here. **Bonus points for messy desks or interesting stuff in the background.**

### Working with NPM

This is a few helpful notes for all of us who do not regularly work with the node/NPM ecosystem and can't remember how to do stuff.

#### Bumping and releasing a new version

    $ make version [major | minor | patch]
    $ make publish


CHANGELOG.md
1.0.0
-----

* Stream ID no longer required to start; instead, use first stream.
* Added Stream list to side of dash; use keys / mouse to switch streams.
* Updated all deps (some were very out of date).
* Main loop refactored to promises.
* Simplified some internal APIs.
* Removed need for credentials file (all arguments may be passed via args)
* Added `--help` option.
* Added `--cred-file-path` option.
* Credentials file may now contain any option.
* Added `--poll-interval` option (default `1000`).
* Renamed `--host` to `--server-url`.
* The user is now prompted for missing mandatory options, including password.




lib/menu.js
'use strict';

const createMenu = require('@timothygu/terminal-menu');
const api = require('./graylog-api');
const Promise = require('bluebird');

module.exports = function makeMenu(options) {
  return api.streams(options)
  .then((streams) => {
    const menu = createMenu({ width: 120, x: 4, y: 2, fg: 'black' });
    menu.reset();
    menu.write('CHOOSE A GRAYLOG STREAM\n');
    menu.write('-----------------------\n');

    // Add items.
    streams
    .sort((s1, s2) => s1.title.localeCompare(s2.title))
    .forEach((s) => {
      menu.add(s.title);
    });
    menu.write('\n');
    menu.add('EXIT');

    // Bind input.
    process.stdin.pipe(menu.createStream()).pipe(process.stdout);
    process.stdin.setRawMode(true);
    menu.on('close', function () {
      process.stdin.setRawMode(false);
      process.stdin.end();
    });

    // When the user selects an option, resolve a promise with the label of that option.
    return Promise.fromNode((callback) => {
      menu.on('select', (label) => callback(null, label));
    })
    .then((label) => {
      menu.close();
      if (label === 'EXIT') {
        console.log('No stream selected; exiting.');
        return process.exit(0);
      }
      // Find the stream matching this title. We could potentially have an issue if two streams
      // have the same title.
      const thisStream = streams.find((s) => s.title === label);
      if (!thisStream) return makeMenu(options); // try again
      return thisStream.id;
    });
  });
};


lib/screen.js
var blessed = require("blessed");
/**
 * This file is part of Graylog2.
 *
 * Graylog2 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Graylog2 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Graylog2.  If not, see <http://www.gnu.org/licenses/>.
 */


const contrib = require("blessed-contrib");

const positions = {
  // [row, col, rowSpan, colSpan]
  totalThroughput:  [0, 2, 1, 2],
  alerts:           [1, 2, 1, 2],
  streamThroughput: [0, 0, 2, 2],
  messageList:      [2, 0, 2, 4],
  streamsList:      [0, 4, 4, 1],
};
const widgets = {};

module.exports = {

  create(options) {
    if (module.exports.screen) throw new Error('Screen already created!');

    const screen = module.exports.screen = blessed.screen({
      autoPadding: true,
      smartCSR: true,
    });
    const grid = module.exports.grid = new contrib.grid({rows: 4, cols: 5, screen});

    // Set up widgets.
    widgets.streamThroughput = grid.set(...positions.streamThroughput, contrib.line, {
      label: "Stream Throughput (max last 5 minutes)",
      style: {
        line: "red",
        text: "green",
        baseline: "white"
      },
      showNthLabel: 60
    });

    widgets.totalThroughput = grid.set(...positions.totalThroughput, contrib.line, {
      label: "Total Throughput (max last 5 minutes)",
      style: {
        line: "green",
        text: "green",
        baseline: "white"
      },
      showNthLabel: 60
    });

    widgets.alerts = grid.set(...positions.alerts, blessed.list, {
      label: "Alerts (30s cached)",
      tags: true,
      items: ["Loading ..."],
      mouse: true,
      scrollable: true
    });

    widgets.messageList = grid.set(...positions.messageList, contrib.log, {
      fg: "green",
      selectedFg: "green",
      bufferLength: 50,
      label: "Messages (Throughput: 0/sec)"
    });

    widgets.streamsList = grid.set(...positions.streamsList, blessed.list, {
      keys: true,
      fg: 'white',
      selectedFg: 'white',
      selectedBg: 'blue',
      interactive: true,
      label: 'Available Streams',
      mouse: true,
      border: {type: "line", fg: "cyan"},
      style: {
        selected: {
          bg: 'green',
          bold: true,
          underline: true,
        }
      }
    });

    widgets.streamsList.on('select', options.onStreamChange);

    screen.render();

    screen.key(['escape', 'q', 'C-c'], function(ch, key) {
      return process.exit(0);
    });
  },

  flush(streamName, streams) {
    Object.keys(widgets).forEach((key) => {
      const w = widgets[key];
      if (w === widgets.streamsList) {
        const titles = streams.map((s) => s.title);
        w.setItems(titles);
        w.select(titles.indexOf(streamName));
      } else {
        if (w.setData) w.setData([{x: 0, y: 0}]);
        if (w.clearItems) w.clearItems();
      }
      if (w === widgets.messageList || w === widgets.streamThroughput || w === widgets.alerts) {
        const {label} = w.options;
        label.replace(/ \(Stream: .*\)/, '');
        w.setLabel(`${label} (Stream: ${streamName})`);
      }
    });
    module.exports.render();
  },

  getWidget(name) {
    if (widgets[name]) return widgets[name];
    throw new Error(`Invalid widget name: ${name}. Available widgets: ${Object.keys(widgets)}`);
  },

	render() {
		module.exports.screen.render();
	},
};


lib/handlers.js
/**
 * This file is part of Graylog2.
 *
 * Graylog2 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Graylog2 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Graylog2.  If not, see <http://www.gnu.org/licenses/>.
 */


const moment = require('moment');
const ui = require('./screen');

const totalThroughputData = [];
const streamThroughputData = [];

module.exports = {

	updateMessagesList(messages) {
		const messageList = ui.getWidget('messageList');

		const items = messages
		.sort((message1, message2) => new Date(message1.timestamp) - new Date(message2.timestamp))
		.map((message) => `${moment(message.timestamp).format()} - ${message.message}`);

		messageList.setItems(items);
	},

	updateTotalThroughputLine(throughput) {
		const throughputLine = ui.getWidget('totalThroughput');

		// Manage local data array.
		if(totalThroughputData.length >= 300) {
			totalThroughputData.shift();
		}
		totalThroughputData.push({x: moment().format("HH:mm"), y: throughput});

		throughputLine.setData(transformLineData(totalThroughputData));
	},

	updateStreamThroughput(throughput) {
		if (!throughput) throughput = 0;
		// Update in title of log messages widget.
		const messageList = ui.getWidget('messageList');
		const label = getWidgetLabel(messageList);
		messageList.setLabel(label.replace(/(Throughput: )(\d+)/, "$1" + throughput));

		// Update in chart.
		const throughputLine = ui.getWidget('streamThroughput');

		// Manage local data array.
		while(streamThroughputData.length >= 300) {
			streamThroughputData.shift();
		}
		streamThroughputData.push({x: moment().format("HH:mm"), y: throughput});

		throughputLine.setData(transformLineData(streamThroughputData));
	},

	updateStreamsList(streams) {
		const streamsList = ui.getWidget('streamsList');
		streamsList.setItems(streams.map((s) => s.title));
		streamsList.focus();
	},

	renderAlerts(alertList) {
		const alerts = ui.getWidget('alerts');
		const lines = [];

		if(alertList.total_triggered == 0) {
			lines.push(""); // Empty line as padding.
			lines.push("{center}{green-fg}No active alerts for this stream!{/green-fg}{/center}");
		} else {
			// Stream has active alerts!
			let msg;
			if (alertList.total_triggered == 1) {
				msg = "One active stream alert:";
			} else {
				msg = "Multiple (" + alertList.total_triggered + ") active stream alerts:";
			}

			lines.push(`{center}{red-bg}!! ${msg} !!{/red-bg}{/center}`);
			lines.push("");

			// Add alerts to list.
			for(const i in alertList.results) {
				lines.push(buildAlertDescription(i, alertList.results[i]));
			}

			lines.push("");
			lines.push("{center}Open your Graylog web interface for alert details.{/center}");
		}

		alerts.setItems(lines);
	}
};

// Blessed-contrib expects an object with x/y array datapoints.
function transformLineData(dataArray) {
	return dataArray.reduce((memo, d) => {
		memo.x.push(d.x);
		memo.y.push(d.y);
		return memo;
	}, {x: [], y: []});
}

function buildAlertDescription(i, alert) {
	return `{red-fg}#${i + 1}:{/red-fg} ${alert.condition.type} alert. (Condition created by ${alert.condition.creator_user_id})`;
}

// We can't just get the label from options, it'll be stale.
function getWidgetLabel(widget) {
	return widget.children.find((c) => c._isLabel).content;
}


lib/controller.js
var graylog = require("./graylog-api.js")
var ui = require("./screen.js")
var handlers = require("./handlers.js")
var ticker = {
	msg: 0,
	stream: 0,
	total: 0,
	alert: 0
}

module.exports = {
	config: {},
	setupHandler: function() {
		var _this=this
		ticker.msg = setInterval(function() {
			graylog.lastMessagesOfStream({
				serverUrl: _this.config.serverUrl,
				streamId: _this.config.streamId,
				mainQuery: _this.config.mainq,
				username: _this.config.apiUser,
				password: _this.config.apiPass
			}, handlers.updateMessagesList)
		}, 1000)
		ticker.stream = setInterval(function() {
			graylog.streamThroughput({
				serverUrl: _this.config.serverUrl,
				streamId: _this.config.streamId,
				username: _this.config.apiUser,
				password: _this.config.apiPass
			}, handlers.updateStreamThroughput)
		}, 1000)
		ticker.total = setInterval(function() {
			graylog.totalThroughput({
				serverUrl: _this.config.serverUrl,
				username: _this.config.apiUser,
				password: _this.config.apiPass
			}, handlers.updateTotalThroughputLine)
		}, 1000)
		ticker.alert = setInterval(function() {
			graylog.streamAlerts({
				serverUrl: _this.config.serverUrl,
				streamId: _this.config.streamId,
				username: _this.config.apiUser,
				password: _this.config.apiPass
			}, handlers.renderAlerts)
		}, 1000)
	},
	slaveHandler: function() {
		var _this=this

		ticker.msg = setInterval(function() {
			graylog.lastMessagesOfStream({
				serverUrl: _this.config.serverUrl,
				streamId: _this.config.streamId,
				mainQuery: _this.config.mainq,
				username: _this.config.apiUser,
				password: _this.config.apiPass
			}, handlers.updateMessagesList)
		}, 1000)
		ticker.stream = setInterval(function() {
			graylog.lastMessagesOfStream({
				serverUrl: _this.config.serverUrl,
				streamId: _this.config.slaveStream,
				username: _this.config.apiUser,
				password: _this.config.apiPass
			}, handlers.updateSlaveList)
		}, 1000)
	},
	resetHandler: function() {
		for (var i in ticker) {
			clearInterval(i);
		}
	}
}

lib/graylog-api.js
/**
 * This file is part of Graylog2.
 *
 * Graylog2 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Graylog2 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Graylog2.  If not, see <http://www.gnu.org/licenses/>.
 */


const Promise = require('bluebird');
const request = require('request');
Promise.promisifyAll(request);

module.exports = {

  streamAlerts(options) {
    const url = `${options.serverURL}streams/${options.streamID}/alerts/check`;

    return makeRequest({uri: url, json: true, auth: makeAuth(options)})
    .then((response) => {
      return response.body;
    })
    .catch((error) => {
      throw new Error('Could not get stream alerts. Error:' + error.message);
    });
  },

  lastMessagesOfStream(options) {
    const url = `${options.serverURL}search/universal/relative`;
    const parameters = {
      query: "*",
      range: 86400,
      fields: "timestamp,message",
      filter: "streams:" + options.streamID,
      limit: 50
    };

    return makeRequest({uri: url, json: true, qs: parameters, auth: makeAuth(options)})
    .then((response) => {
      return response.body.messages.map((m) => m.message);
    })
    .catch((error) => {
      throw new Error('Could not get stream messages. ' + error.message);
    });
  },

  streams(options) {
    const url = `${options.serverURL}streams`;

    return makeRequest({uri: url, json: true, auth: makeAuth(options)})
    .then((response) => {
      return response.body.streams
      .sort((a, b) => a.title.localeCompare(b.title));
    })
    .catch((error) => {
      throw new Error('Could not get streams. ' + error.stack);
    });
  },

  streamThroughput(options) {
    const url = `${options.serverURL}system/metrics/org.graylog2.plugin.streams.Stream.${options.streamID}.incomingMessages.1-sec-rate`;

    return makeRequest({uri: url, json: true, auth: makeAuth(options)})
    .then((response) => {
      return response.body.value;
    })
    .catch((error) => {
      throw new Error('Could not get stream throughput. ' + error.message);
    });
  },

  totalThroughput(options) {
    const url = `${options.serverURL}system/throughput`;

    return makeRequest({uri: url, json: true, auth: makeAuth(options)})
    .then((response) => {
      return response.body.throughput;
    })
    .catch((error) => {
      throw new Error('Could not get stream total throughput. ' + error.message);
    });
  }
};

function makeRequest(options) {
  return request.getAsync(options)
  .then((response) => {
    if (response.statusCode < 200 || response.statusCode >= 400) {
      throw new Error(`Bad status code: ${response.statusCode} ${response.statusMessage} ${response.body || ''}`);
    }
    return response;
  });
}

// TODO API Token?
function makeAuth(options) {
  return {user: options.username, pass: options.password, sendImmediately: true};
}


graylog2-dashboard.js
/**
 * This file is part of Graylog2.
 *
 * Graylog2 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Graylog2 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Graylog2.  If not, see <http://www.gnu.org/licenses/>.
 */

var graylog = require("./lib/graylog-api.js")
var ui = require("./lib/screen.js")
var handlers = require("./lib/handlers.js")

/*
 *
 *   ┈╱╱▏┈┈╱╱╱╱▏╱╱▏┈┈┈    ┈╱╱▏┈┈╱╱╱╱▏╱╱▏┈┈┈    ┈╱╱▏┈┈╱╱╱╱▏╱╱▏┈┈┈
 *   ┈▇╱▏┈┈▇▇▇╱▏▇╱▏┈┈┈    ┈▇╱▏┈┈▇▇▇╱▏▇╱▏┈┈┈    ┈▇╱▏┈┈▇▇▇╱▏▇╱▏┈┈┈
 *   ┈▇╱▏▁┈▇╱▇╱▏▇╱▏▁┈┈    ┈▇╱▏▁┈▇╱▇╱▏▇╱▏▁┈┈    ┈▇╱▏▁┈▇╱▇╱▏▇╱▏▁┈┈
 *   ┈▇╱╱╱▏▇╱▇╱▏▇╱╱╱▏┈    ┈▇╱╱╱▏▇╱▇╱▏▇╱╱╱▏┈    ┈▇╱╱╱▏▇╱▇╱▏▇╱╱╱▏┈
 *   ┈▇▇▇╱┈▇▇▇╱┈▇▇▇╱┈┈    ┈▇▇▇╱┈▇▇▇╱┈▇▇▇╱┈┈    ┈▇▇▇╱┈▇▇▇╱┈▇▇▇╱┈┈
 *
 *   I have no clue about JavaScript or even node.js and this is 
 *   going to be pretty terrible code. It is a wonder that I got
 *   it running at all lol.
 *
 *   I wish this was Java.
 *
 *   Have fun in here! (Lennart, 01/2015)
 *
 */

setInterval(function() {
  graylog.lastMessagesOfStream({
    serverUrl: "http://localhost:12900",
    streamId: "549d7f9fbee84e568d181655",
    username: "lennart",
    password: "123123123"
  }, handlers.updateMessagesList)
}, 1000)

setInterval(function() {
  graylog.streamThroughput({
    serverUrl: "http://localhost:12900",
    streamId: "549d7f9fbee84e568d181655",
    username: "lennart",
    password: "123123123"
  }, handlers.updateStreamThroughput)
}, 1000)

setInterval(function() {
  graylog.totalThroughput({
    serverUrl: "http://localhost:12900",
    username: "lennart",
    password: "123123123"
  }, handlers.updateTotalThroughputLine)
}, 1000)

graylog-dashboard.js
#!/usr/bin/env node
'use strict';

/**
 * This file is part of Graylog2.
 *
 * Graylog2 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Graylog2 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Graylog2.  If not, see <http://www.gnu.org/licenses/>.
 */

const Promise = require('bluebird');
const graylog = require('./lib/graylog-api');
const ui = require('./lib/screen');
const handlers = require('./lib/handlers');
const fs = require('fs');
const yaml = require('js-yaml');
const inquirer = require('inquirer');
const yargs = require('yargs')
  .usage('Usage: $0 <command> [options]')
  .describe('stream-title', 'Graylog Stream Title')
  .describe('server-url', 'Full Graylog REST API URL')
  .describe('poll-interval', 'How often (in ms) to poll the Graylog server')
  .default('poll-interval', 1000)
  .describe('username', 'Graylog username')
  .describe('password', 'Graylog password')
  .describe('cred-file-path', 'Path to an optional credentials file')
  .default('cred-file-path', `${process.env.HOME}/.graylog_dashboard`)
  .help();

// Will contain:
// username, password, host, streamID, pollInterval, credFilePath
let config;

function getOptions() {
  return Promise.try(function getOptionInput() {
    const {argv} = yargs;

    // Read user credentials.
    let fileConfig = {};
    try {
      fileConfig = yaml.safeLoad(fs.readFileSync(argv['cred-file-path'], 'utf8'));
    } catch (err) {
      // Ignore, maybe everything was defined in argv
    }

    // Remove falsy values from argv to clean up merge.
    Object.keys(argv).forEach((k) => {
      if (!argv[k]) delete argv[k];
    });
    // Merge argv & file config.
    config = Object.assign({}, fileConfig, argv);
    config.serverURL = argv['server-url'] || config.serverURL; // casing
  })
  .then(function checkMissingOptions() {
    // Check mandatory config.
    const shouldPrompt = ['username', 'password', 'serverURL'].filter((k) => !config[k]);

    // Prompt user for missing options.
    if (shouldPrompt.length) {
      const questions = shouldPrompt.map((k) => (
        {name: k, message: `Please input ${k}:`, type: k === 'password' ? 'password' : 'input'}
      ));

      // Inquirer uses core `readline`. When activated, it automatically emits `line` and `keypress` events.
      // Built-in NodeJS `readline` uses `stream[Symbol('keypress-decoder')]` to mark this, but blessed
      // uses its own check. If both are running, we'll get double keypress events which makes working
      // the stream list very difficult.
      process.stdin._keypressDecoder = true;

      const prompt = inquirer.createPromptModule();
      return prompt(questions)
      .then(function(result) {
        Object.assign(config, result);
      });
    }
  })
  .then(function coerceOptions() {
    // Make sure we have a protocol (default: https)
    if (config.serverURL.slice(0, 4) !== 'http') config.serverURL = 'https://' + config.serverURL;
    // Make sure we have a port (default REST API port is 12900)
    if (!/:\d+$/.test(config.serverURL)) config.serverURL += ':12900';
    // Make sure config.serverURL has a trailing slash. (computers.)
    if (config.serverURL[config.serverURL.length - 1] !== '/') config.serverURL += '/';
    return config;
  });
}

let streams; // filled in after streams runs
function storeStreams(_streams) {
  streams = _streams;
  if (!config.streamTitle) {
    // Could happen on startup
    config.streamTitle = streams[0].title;
  }
  const thisStream = streams.find((s) => s.title === config.streamTitle);
  config.streamID = thisStream.id;
  ui.flush(thisStream.title, streams);
}

function onStreamChange(blessedEl) {
  const streamTitle = blessedEl.content;
  config.streamTitle = streamTitle;
  config.streamID = streams.find((s) => s.title === streamTitle).id;
  ui.flush(streamTitle, streams);
}

function poll() {
  return Promise.all([
    graylog.lastMessagesOfStream(config).then(handlers.updateMessagesList),
    graylog.streamThroughput(config).then(handlers.updateStreamThroughput),
    graylog.totalThroughput(config).then(handlers.updateTotalThroughputLine),
    graylog.streamAlerts(config).then(handlers.renderAlerts),
  ])
  .then(() => ui.render()) // batch up renders
  .delay(config.pollInterval)
  .then(() => poll(config.streamID));
}

// Entry
getOptions()
.then(() => ui.create({onStreamChange}))
.then(() => graylog.streams(config))
.tap(storeStreams)
.then(handlers.updateStreamsList)
.then(poll)
.catch((e) => { ui.screen && ui.screen.destroy(); console.error(e); process.exit(1); });
